Use env var in spark container (#1522)
apache/polaris: site/content/in-dev/unreleased/getting-started/using-polaris.md

type: docs
weight: 400
---
## Before start

> Put your `CLIENT_ID` & `CLIENT_SECRET` in .env file for later use.

Reviewer: This is literally the first line in the quickstart guide. I don't know what these things are.

Author: Added more explicit instruction.

> Define your `CLIENT_ID` & `CLIENT_SECRET` and put them in .env file for later use.

Reviewer: Wait sorry, I just realized that this is already detailed in the [quickstart](https://polaris.apache.org/in-dev/unreleased/getting-started/quickstart/) page which links into here:

```
When Polaris is run using the ./gradlew run command, the root principal credentials are root and secret for the CLIENT_ID and CLIENT_SECRET, respectively. You can also set these credentials as environment variables for use with the Polaris CLI:

export CLIENT_ID=root
export CLIENT_SECRET=secret

```

So we now have redunandant but inconsistent instructions when going from the quickstart into `using-polaris`

Reviewer: Can't we just keep suggesting that users `export` rather than echo things into `.env` here?

.env is useful for when we're trying to automate things, but it seems overcomplicated to have to do all this just to set an env var:
```
echo "USER_CLIENT_ID=XXXX" >> .env
echo "USER_CLIENT_SECRET=YYYY" >> .env
cd polaris
set -a
source .env
cd spark
```

Author: I don't think this is conflicting.
Using Polaris is another page.

Author: > Can't we just keep suggesting that users `export` rather than echo things into `.env` here?
> 
> .env is useful for when we're trying to automate things, but it seems overcomplicated to have to do all this just to set an env var:
> 
> ```
> echo "USER_CLIENT_ID=XXXX" >> .env
> echo "USER_CLIENT_SECRET=YYYY" >> .env
> cd polaris
> set -a
> source .env
> cd spark
> ```

I originally is use export, so you prefer changing it back?

Reviewer: I think export makes sense for places where we want the user to manually manage credentials within a session so that they can be used later on. .env makes sense if we are automating something and need to manage credentials on the user's behalf.

Be sure to provide the necessary credentials, hostname, and port as before.

When the `principals create` command completes successfully, it will return the credentials for this new principal. Add the credentials in .env for later. For example:

```
./polaris ... principals create example
{"clientId": "XXXX", "clientSecret": "YYYY"}
echo "USER_CLIENT_ID="XXXX"                                                

> USER_CLIENT_SECRET="YYYY"" >> .env

Author: Thanks, just fixed them.

--conf spark.sql.catalog.quickstart_catalog=org.apache.iceberg.spark.SparkCatalog \
--conf spark.sql.catalog.quickstart_catalog.catalog-impl=org.apache.iceberg.rest.RESTCatalog \
--conf spark.sql.catalog.quickstart_catalog.uri=http://localhost:8181/api/catalog \
--conf spark.sql.catalog.quickstart_catalog.credential='${USER_CLIENT_ID}:${USER_CLIENT_SECRET}' \
--conf spark.sql.catalog.quickstart_catalog.scope='PRINCIPAL_ROLE:ALL' \
--conf spark.sql.catalog.quickstart_catalog.token-refresh-enabled=true \
--conf spark.sql.catalog.quickstart_catalog.client.region=us-west-2
```

Similar to the CLI commands above, this configures Spark to use the Polaris running at `localhost:8181`. If your Polaris server is running elsewhere, but sure to update the configuration appropriately.

Finally, note that we include the `hadoop-aws` package here. If your table is using a different filesystem, be sure to include the appropriate dependency.

> #### Using Spark SQL from a Docker container

Reviewer 2: ```suggestion
#### Using Spark SQL from a Docker container

Refresh the Docker container with the user's credentials:
```

Replace the credentials used in the Docker container using the following code:

```shell

> export USER_CLIENT_ID="XXXX"

Reviewer 3: @eric-maynard 
Maybe use `.env` instead?

Author: I thought this is a demonstration page? Isn't this a one-time thing?

Reviewer 3: Fair enough. What I was thinking is, it will be 2 less lines to copy/paste and we put some default credential such as `root:s3cr3t` in the `.env` and add a note said to use a diff credential, update `.env` etc.

Reviewer: .env is a good idea, I wish I would have thought of that!

@owenowenisme you're right that this is a one-time thing... but that's also why modifying the `getting-started/eclipselink/docker-compose.yml` is suspect. 

Maybe we could write to a new file and `source` it instead of using the normal .env? Or we could just append to the end of the existing .env.

Author: Yeah we could use `.env` file.
And I'll use `SPARK_USER_CLIENT_ID` `SPARK_USER_CLIENT_SECRET` as varName.

WDYT?

Reviewer 2: I'd just use `USER_CLIENT_ID` and `USER_CLIENT_ID` - and we should extend this change to Trino below as well. We're doing the same thing there.

Reviewer 2: If we are doing `.env` - let's also make sure to add it to the `.gitignore` or else we land into the same problem of users might accidentally commit their credentials using that file instead 😅

> Replace the credentials used in the Docker container using the following code:

Reviewer 2: ```suggestion
Refresh the Docker container with the user's credentials:
```
